### PR TITLE
feat: resolve flags on context change

### PR DESCRIPTION
**Pull request by andreas-karlsson**
## Hi There, I just made a Pull Request!
Mostly changes that make our OpenFeature web-provider re-resolve whenever the associated Confidence instance updates it's context.

#### :heavy_check_mark: Checklist

- [ ] All tests are passing
- [ ] Relevant documentation updated
- [ ] linter/style run on changed files
- [ ] Tests added for new functionality
- [ ] Regression tests added for bug fixes
- [ ] Tested in a corresponding example app

### CODE REVIEW

File: packages/sdk/src/Confidence.ts [hunk 1]
```diff
@@ -33,10 +34,29 @@ export class Confidence implements EventSender, Trackable {
...
+  private flagResolution?: FlagResolutionPromise;
 
   constructor(config: Configuration, parent?: Confidence) {
     this.config = config;
     this.parent = parent;
+    this.contextChanges = subject(observer => {
+      let parentSubscription: Closer | void;
+      if (parent) {
+        parentSubscription = parent.contextChanges(keys => {
+          observer(keys.filter(key => !this._context.has(key)));
```
Comment (andreas-karlsson) [inline packages/sdk/src/Confidence.ts:51]: If all keys are filtered we shouldn't emit!

File: packages/sdk/src/Confidence.ts [hunk 2]
```diff
@@ -73,17 +93,26 @@ export class Confidence implements EventSender, Trackable {
   }
 
   setContext(context: Context): void {
+    const current = this.getContext();
+    const changedKeys: string[] = [];
     for (const key of Object.keys(context)) {
-      this.updateContextEntry(key, context[key]);
+      if (Value.equal(current[key], context[key])) continue;
```
Comment (andreas-karlsson) [inline packages/sdk/src/Confidence.ts:99]: I think (!) we might be able to do reference equals here.

File: packages/sdk/src/Confidence.ts [hunk 3]
```diff
@@ -73,17 +93,26 @@ export class Confidence implements EventSender, Trackable {
...
+    if (this.contextChanged && changedKeys.length > 0) {
+      this.contextChanged(changedKeys);
     }
-  }
-
-  private updateContextEntry<K extends string>(name: K, value: Context[K]) {
-    this._context.set(name, Value.clone(value));
   }
 
   clearContext(): void {
```
Comment (andreas-karlsson) [inline packages/sdk/src/Confidence.ts:108]: We should remove this function. It's never really useful and it complicates things.

File: packages/sdk/src/Confidence.ts [hunk 4]
```diff
@@ -106,7 +135,15 @@ export class Confidence implements EventSender, Trackable {
    * @internal
    */
   resolve(flagNames: string[]): Promise<FlagResolution> {
-    return this.config.flagResolverClient.resolve(this.getContext(), flagNames);
+    // we first resolve a promise so that if multiple context changes happen in the same tick, we still only make one resolve
+    return Promise.resolve().then(() => {
+      const context = this.getContext();
+      if (!this.flagResolution || !Value.equal(context, this.flagResolution.context)) {
+        this.flagResolution?.abort();
```
Comment (andreas-karlsson) [inline packages/sdk/src/Confidence.ts:142]: Should put a reasonable reason in the abort call. Now it's just a generic "opeation was aborted" or something similar.

File: packages/sdk/src/context.ts [hunk 5]
```diff
@@ -9,9 +9,7 @@ export interface Contextual<Self extends Contextual<Self>> {
 
 export interface Context extends Value.Struct {
   visitor_id?: string;
-  openFeature?: Value.Struct & {
-    targeting_key?: string;
-  };
+  targeting_key?: string;
```
Comment (andreas-karlsson) [inline packages/sdk/src/context.ts:12]: Do we even want semantic context for targeting_key? Do we ever expect that to be a thing in Confidence?

File: packages/sdk/src/observing.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        pendingEmit = timeout(emit);
+      }
+      for (const value of values) {
+        buffer.add(value);
+      }
+    });
+    const closeEmit = () => pendingEmit?.();
+    return Closer.combine(closeEmit, closeSource);
+  };
+}
```
Comment (andreas-karlsson) [inline packages/sdk/src/observing.ts:75]: We don't use this any longer, so could be ditched? But might become useful in the future...?

File: packages/sdk/src/Confidence.ts (hunk 2 above)
Comment (andreas-karlsson) [reply packages/sdk/src/Confidence.ts:100]: Or actually we won't. There's the case where the child set's the same value as the parent, that would trigger a spurious context change. Added a test for it.

File: packages/sdk/src/Confidence.ts (hunk 3 above)
Comment (andreas-karlsson) [reply packages/sdk/src/Confidence.ts:109]: Maybe later...

File: packages/sdk/src/context.ts (hunk 5 above)
Comment (andreas-karlsson) [reply packages/sdk/src/context.ts:12]: For later...

File: packages/sdk/src/observing.ts (hunk 6 above)
Comment (andreas-karlsson) [reply packages/sdk/src/observing.ts:75]: Will remove!

**Review by nicklasl — APPROVED** [review]

**Pull request merged**